### PR TITLE
[WIP] Add singlefile push run

### PR DESCRIPTION
**Pull request by wprater**
A WIP to push a single file.

- you cannot push a single directory
- push a single file should follow project path

not currently writing to the remote store, but wanted to get this up if it's helpful for #272  as I may be offline for a couple days.

### CODE REVIEW

File: lib/tessel/deploy.js [hunk 1]
```diff
@@ -100,7 +108,7 @@ actions.execRemoteCommand = function(tessel, command, filepath) {
   });
 };
 
-actions.findProject = function(file) {
+actions.findProject = function(file, single) {
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: Change this to: 

```js
actions.findProject = function(file, options) {
```

File: lib/tessel/deploy.js [hunk 2]
```diff
@@ -177,16 +189,25 @@ actions.sendBundle = function(t, filepath, opts) {
       .then(function(remoteProcess) {
         var error = '';
 
-        actions.findProject(opts.entryPoint).then(function(project) {
-          var resolved = path.resolve(process.cwd(), project.pushdir);
+        actions.findProject(opts.entryPoint, opts.single).then(function(project) {
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: Change to: 

```js
actions.findProject(opts)
```

**Comment by rwaldron** [comment]
In addtion to the inline review notes, please review the travis-ci failures here: https://travis-ci.org/tessel/t2-cli/builds/80551104 (be sure to always run `grunt`)

File: lib/tessel/deploy.js (hunk 1 above)
Comment (wprater) [inline lib/tessel/deploy.js]: good call, I went with `opts` since that's what been used in the rest of the file.

File: bin/tessel-2.js [hunk 3]
```diff
@@ -124,6 +129,11 @@ makeCommand('push')
     required: true,
     help: 'The entry point file to deploy to Tessel'
   })
+  .option('single', {
+    flag: true,
+    abbr: 's',
+    help: 'Push only the entryPoint'
```
Comment (rwaldron) [inline bin/tessel-2.js]: `entryPoint` => `entry point`

File: bin/tessel-2.js [hunk 4]
```diff
@@ -107,6 +107,11 @@ makeCommand('run')
...
```
Comment (rwaldron) [inline bin/tessel-2.js]: `Push only the entryPoint` => `Run only the specified entry point file.`

File: lib/tessel/deploy.js [hunk 5]
```diff
@@ -100,8 +108,10 @@ actions.execRemoteCommand = function(tessel, command, filepath) {
   });
 };
 
-actions.findProject = function(file) {
+actions.findProject = function(opts) {
   return new Promise(function(resolve, reject) {
+    var single = opts.single;
+    var file = opts.file;
```
Comment (rwaldron) [inline lib/tessel/deploy.js]: `var file = opts.file;` => `var file = opts.entryPoint;`

Comment (wprater) [inline lib/tessel/deploy.js]: durr, sorry.  I need to write test for this, but could use help to get myself situated with the testing stubs and methodology we're using.

**Comment by rwaldron** [comment]
To answer your question in the comment, regarding tests: take a look at https://github.com/tessel/t2-cli/blob/master/test/unit/deploy.js to get started.

- Create new tests after this block: https://github.com/tessel/t2-cli/blob/master/test/unit/deploy.js#L81-L114 that _almost_ replicate those two, however ensuring that the `deleteFolder` remote command is not called.
- Using the fixtures in test/unit/fixtures, capture the size of a deployment in the "whole project" case, then compare that to single file run/push (there might be better ways to test this)

If I think of more I will post here. I will be back at my regular work area tomorrow and can be more useful then.

**Comment by wprater** [comment]
@rwaldron thanks.  currently a single file does not seem to be writing [here](https://github.com/wprater/t2-cli/blob/add-singlefile-push-run/lib/tessel/deploy.js#L235). will look to tests in the next couple days.

**Comment by rwaldron** [comment]
> currently a single file does not seem to be writing here. will look to tests in the next couple days.

I'll also take a look

**Comment by rwaldron** [comment]
The problem with the current solution is that it's trying to send a buffer containing the contents of a file, but the other end expects a buffer for a tarball. The solution doesn't require two separate reading paths, because fstream-ignore can have new rules added dynamically. I will make a pull request to your branch with the latest changes.

**Comment by rwaldron** [comment]
Actually, it's easier to land this and then merge the modification.

**Comment by wprater** [comment]
OK.. I can make some updates now.. or I can rebase this into one WIP commit?

**Comment by wprater** [comment]
we can just have it use a new command for the single file vs the `untarStdin`?

**Comment by rwaldron** [comment]
> we can just have it use a new command for the single file vs the untarStdin?

Then we have to maintain two code paths that effectively do the same thing.

**Comment by rwaldron** [comment]
> OK.. I can make some updates now.. or I can rebase this into one WIP commit?

Sorry, I missed this after my last comment. I'd already landed yours locally and was working on the tests when this came through.

**Comment by wprater** [comment]
>> we can just have it use a new command for the single file vs the untarStdin?

> Then we have to maintain two code paths that effectively do the same thing.

OK.  will check out what you come up with! thanks

**Comment by rwaldron** [comment]
We still need tests that ensure that the `deleteFolder` command is not called when `tessel.deployScript({ single: true, ... })`

**Comment by wprater** [comment]
will add that test.

**Pull request merged**